ttnn/cpp/ttnn/graph_processor.cpp

```diff
@@ -0,0 +1,503 @@
+        if (tensor.has_value()) {
+            int tensor_id = add_tensor(tensor.value());
+            graph[tensor_id].connections.push_back(current_op_id.top());
+        }
+    }
+    void GraphProcessor::end_op_process_vector(const std::any& any_val) {
+        const auto& tensor_vec = std::any_cast<std::reference_wrapper<std::vector<Tensor>>>(any_val).get();
+        for (int j = 0; auto& it : tensor_vec) {
+            int tensor_id = add_tensor(it);
+            graph[last_finished_op_id].connections.push_back(tensor_id);
+            j++;
+        }
+    }
+    void GraphProcessor::end_op_process_vector_optional(const std::any& any_val) {
+        const auto& tensor_vec = std::any_cast<std::reference_wrapper<std::vector<std::optional<Tensor>>>>(any_val).get();
+        for (int j = 0; auto& it : tensor_vec) {
+            if (it.has_value()) {
+                int tensor_id = add_tensor(it.value());
+                graph[last_finished_op_id].connections.push_back(tensor_id);
+                j++;
+            }
+        }
+    }
+    void GraphProcessor::end_op_process_vector_optional_const(const std::any& any_val) {
+        const auto& tensor_vec = std::any_cast<std::reference_wrapper<std::vector<std::optional<const Tensor>>>>(any_val).get();
+        for (int j = 0; auto& it : tensor_vec) {
+            if (it.has_value()) {
+                int tensor_id = add_tensor(it.value());
+                graph[last_finished_op_id].connections.push_back(tensor_id);
+            }
+        }
+    }
+    void GraphProcessor::end_op_process_tensor(const std::any& any_val) {
+        const auto& tensor = std::any_cast<std::reference_wrapper<Tensor>>(any_val).get();
+        int tensor_id = add_tensor(tensor);
+        graph[last_finished_op_id].connections.push_back(tensor_id);
+    }
+    void GraphProcessor::end_op_process_optional_tensor(const std::any& any_val) {
+        const auto& tensor = std::any_cast<std::reference_wrapper<std::optional<Tensor>>>(any_val).get();
+        if (tensor.has_value()) {
+            int tensor_id = add_tensor(tensor.value());
+            graph[last_finished_op_id].connections.push_back(tensor_id);
+        }
+    }
+
+    void GraphProcessor::begin_capture(RunMode mode) {
+        const std::lock_guard<std::mutex> lock(mutex);
+        graph.clear();
+        id_to_counter.clear();
+        graph.push_back(Vertex{
+            .counter = 0,
+            .name = "capture_start",
+            .params = {},
+            .connections = {}
+        });
+        hooks = std::make_shared<ProcessorHooks>();
+        tt::tt_metal::GraphTracker::instance().add_hook(hooks);
+        hooks->set_block(mode == RunMode::FAKE);
+        current_op_id.push(0);
+    }
+    std::string GraphProcessor::end_capture() {
+        const std::lock_guard<std::mutex> lock(mutex);
+        auto json_object = to_json(graph);
+        return json_object.dump();
```

Comment:
why not just return json out directly?

Reply:
I need to doublecheck but as I remember I tried probably nlohman::json wasnot supported and I got an error.

Reply:
I updated the pybind to convert to python objects (dict/list/bool/int/float/string/None)